toolbox/library/src/YarpRpc.cpp

```diff
@@ -0,0 +1,257 @@
+        return false;
+    }
+
+    // CLASS INITIALIZATION
+    // ====================
+
+    pImpl->network = std::make_unique<yarp::os::Network>();
+
+    if (!yarp::os::Network::initialized() || !yarp::os::Network::checkNetwork(5.0)) {
+        bfError << "YARP server wasn't found active.";
+        return false;
+    }
+
+    // Open a rpc client port
+    pImpl->clientPortName = "...";
+
+    if (!pImpl->rpcClientPort.open(pImpl->clientPortName)) {
+        bfError << "Error while opening yarp rpc server port.";
+        return false;
+    }
+
+    if (!yarp::os::Network::connect(pImpl->rpcClientPort.getName(), pImpl->serverPortName)) {
+        bfError << "Failed to connect " << pImpl->serverPortName << " to "
+                << pImpl->rpcClientPort.getName() << ".";
+        return false;
+    }
+
+    // Store rpc command in bottle
+    pImpl->rpcCommand.fromString(command);
+
+    return true;
+}
+
+bool YarpRpc::terminate(const BlockInformation* /*blockInfo*/)
+{
+    if(pImpl->rpcClientPort.isWriting()) {
+        bfWarning << "Interrupting rpc port";
+        pImpl->rpcClientPort.interrupt();
+    }
+
+    yarp::os::Network::disconnect(pImpl->rpcClientPort.getName(), pImpl->serverPortName);
+
+    // Close the port
+    pImpl->rpcClientPort.close();
+
+    return true;
+}
+
+bool YarpRpc::output(const BlockInformation* blockInfo)
+{
+    InputSignalPtr triggerSignal = blockInfo->getInputPortSignal(InputIndex::TriggerSignal);
+
+    if (!triggerSignal) {
+        bfError << "Input signal not valid.";
+        return false;
+    }
+
+    bool rpcStatus = true;
+
+    //TODO: Update to getting bool type after
+    //testing boolean support more thoroughly
+    double trigger = triggerSignal->get<double>(0);
+
+    if (pImpl->risingEdgeTrigger) {
+
+        if (trigger && !pImpl->previousTrigger) {
+
+            rpcStatus = sendRpcCommand();
+
+        }
+
+    }
+    else if (static_cast<bool>(trigger)) {
+
+        rpcStatus = sendRpcCommand();
+
+    }
+
+    //Update previous trigger
+    pImpl->previousTrigger = trigger;
+
+    //TODO: Check if it is useful to forward reponse to an output signal
+    return rpcStatus;
+}
+
+bool YarpRpc::sendRpcCommand()
+{
+    if (pImpl->rpcClientPort.isWriting()) {
+        bfError << "YarpRpc client port is blocked in writing";
+        return false;
+    }
+
+    yarp::os::Bottle response;
+
+    // Initialize the time counter for the timeout
+    const double t0 = yarp::os::SystemClock::nowSystem();
+
+    while(!pImpl->rpcClientPort.write(pImpl->rpcCommand, response)) {
+
+        yarp::os::Time::delay(0.0005);
+        const double now = yarp::os::SystemClock::nowSystem();
+
+        if((now - t0) > pImpl->timeout) {
+            bfError << "Failed to send the rpc command to the server "
+                    << " with in the configured timeout";
+            return false;
+        }
+    }
+
+    bfWarning << response.toString().c_str();
```

Comment:
Leftover?

Reply:
Nope :grinning: left it consciously as this is the only output users see on the matlab terminal to see if things went fine with the rpc command e.g. `[ok]` is the rpc commands to WBD rpc server. I prefer to leave it for the time being and change it if we decide to handle the rpc response as an output signal from the block at a later stage.

Reply:
Mhh, that is  going to be messy if several RPC blocks are used. How are the users expected to know that the random `[ok]` appearing in the output is the RPC  response? If we need to have this, I would at least suggest to 1) Output as an info, not a warning 2) Have an option to disable it.

Reply:
@traversaro I checked with @diegoferigo and there isn't any info logging in block factory. The warning and error logs are designed to be compatible with the simulink output. So, as at the moment there is not output signal based on the response, we can add optional additional "expected response" parameter to the YarpRpc mask and check the response against it. In case if they do not match, can notify the user directly in simulink using `bferror` logger. For the time being I will leave a `TODO` message and comment out the warning of the response.